src/logworker.cpp

```diff
@@ -120,6 +122,14 @@ namespace g3 {
       return std::unique_ptr<LogWorker>(new LogWorker);
    }
 
+   void LogWorker::removeWrappedSink(std::shared_ptr<g3::internal::SinkWrapper> sink) {
+      auto bg_removesink_call = [this, &sink] {
```

Comment:
@gimesketvirtadieni 

For adding a sink this is not something I worry about at all. 

For removing a sink I understand your point in regard to the example with resource cleanup. Also there is no guarantee that the user hasn't  kept their own shared_ptr of the sink. 

I agree with you that after removing the sink it makes sense to provide a **guarantee** that the sink is done processing all enqueued messages.  That way it does not matter whether or not the `future/packaged_task` keeps it alive for some ms or X time units longer. There should be no more processing of the queue after it is `removed` - even if the actual destruction might be somewhat delayed. 

That's a fairly strong guarantee and I think it would satisfy most users. 


It's an excellent point you bring up to provide a `flush()` function or `synchronization()`/'join()' function. I see no problem adding that. 

If you want that then my suggestion would be to add it to 
https://github.com/KjellKod/g3log/blob/master/src/g3log/sink.hpp

I.e. through it you can use the spawn_task and call `wait()` on an empty lambda.   If for sake of argument we call it 'flush()` then the remove sink could look as follows

```
   void LogWorker::removeWrappedSink(std::shared_ptr<g3::internal::SinkWrapper> sink) {
      auto bg_removesink_call = [this, sink] {
         _impl._sinks.erase(std::remove(_impl._sinks.begin(), _impl._sinks.end(), sink), _impl._sinks.end());
      };
      auto token_done = g3::spawn_task(bg_removesink_call, _impl._bg.get());
      token_done.wait(); // the sink is now guaranteed to not receive any new messages.
      sink->flush(); // the sink is now guaranteed to have processed all messages that were enqueued.
   }
```